crates/core/src/host/wasmtime/wasm_instance_env.rs

```diff
@@ -582,69 +582,121 @@ impl WasmInstanceEnv {
-    /// - advancing the iterator resulted in an error
+    /// The `buffer_len = buffer_len_ptr[..size_of::<usize>()]` stores the capacity of `buffer`.
+    /// On success (`0` or `-1` is returned),
+    /// `buffer_len` is set to the combined length of the encoded rows.
+    /// When `-1` is returned, the iterator has been exhausted
+    /// and there are no more rows to read,
+    /// leading to the iterator being immediately destroyed.
+    /// Note that the host is free to reuse allocations in a pool,
+    /// destroying the handle logically does not entail that memory is necessarily reclaimed.
+    ///
+    /// # Traps
+    ///
+    /// Traps if:
+    ///
+    /// - `buffer_len_ptr` is NULL or `buffer_len` is not in bounds of WASM memory.
+    /// - `buffer_ptr` is NULL or `buffer` is not in bounds of WASM memory.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error:
+    ///
+    /// - `NO_SUCH_ITER`, when `iter` is not a valid iterator.
+    /// - `BUFFER_TOO_SMALL`, when there are rows left but they cannot fit in `buffer`.
+    ///   When this occurs, `buffer_len` is set to the size of the next item in the iterator.
+    ///   To make progress, the caller should reallocate the buffer to at least that size and try again.
     // #[tracing::instrument(skip_all)]
-    pub fn iter_advance(
+    pub fn row_iter_bsatn_advance(
         caller: Caller<'_, Self>,
         iter: u32,
-        buffer: WasmPtr<u8>,
-        buffer_len: WasmPtr<u32>,
-    ) -> RtResult<u32> {
-        let iter = RowIterIdx(iter);
-        Self::cvt(caller, AbiCall::IterNext, |caller| {
+        buffer_ptr: WasmPtr<u8>,
+        buffer_len_ptr: WasmPtr<u32>,
+    ) -> RtResult<i32> {
+        let row_iter_idx = RowIterIdx(iter);
+        Self::cvt_custom(caller, AbiCall::RowIterBsatnAdvance, |caller| {
             let (mem, env) = Self::mem_env(caller);
 
-            // Retrieve the iterator by `iter`.
-            let iter = env.iters.get_mut(iter).context("no such iterator")?;
+            // Retrieve the iterator by `row_iter_idx`, or error.
+            let Some(iter) = env.iters.get_mut(row_iter_idx) else {
+                return Ok(errno::NO_SUCH_ITER.get().into());
+            };
 
-            let buffer_len_ = u32::read_from(mem, buffer_len)?;
-            let mut wasm_buffer = mem.deref_slice_mut(buffer, buffer_len_)?;
+            // Read `buffer_len`, i.e., the capacity of `buffer` pointed to by `buffer_ptr`.
+            let buffer_len = u32::read_from(mem, buffer_len_ptr)?;
+            // Get a mutable view to the `buffer`.
+            let mut buffer = mem.deref_slice_mut(buffer_ptr, buffer_len)?;
 
             let mut written = 0;
-            // Advance the iterator.
+            // Fill the buffer as much as possible.
             while let Some(chunk) = iter.as_slice().first() {
-                let Some(buf_chunk) = wasm_buffer.get_mut(..chunk.len()) else {
+                // TODO(Centril): refactor using `split_at_mut_checked`.
```

Comment:
Alternatively, could leverage the `std::io::Write` implementation on slices. `.write()` on a slice results in the same zero-cost memcpy code and automatically takes care of "advancing" the slice to the remainder part: https://rust.godbolt.org/z/1x4W6KeYx

Reply:
Yeah either way its a future TODO, just wanted to comment with an alternative :)